### PR TITLE
C elegens metadata

### PR DESCRIPTION
**Pull request by bendichter**
## Motivation

From conversation with Daniel Sprague:

> I was wondering if there were possible changes we could make to the restrictions imposed on age and sex. C. elegans growth stages are defined from L1-L4 followed by adulthood. Most of the worms we use are designated 'YA' for young adult. I believe for our purposes, this is a more useful designation than dahs or weeks old since the lifesplan of the owrm is relatively short. Furthermore, for sex, C. elegans have two sexes: male 'XO' and hermaphrodite 'XX' which are designated by their sex chromosomes. I see that there is an option for other, but since the hermaphrodite is the typical system we study, I think it would be useful to be able to just input the chromosomal designations if that's possible.

What do you think, @CodyCBakerPhD ?

### CODE REVIEW

File: src/nwbinspector/checks/nwbfile_metadata.py [hunk 1]
```diff
@@ -197,10 +197,12 @@ def check_subject_id_exists(subject: Subject):
 
 @register_check(importance=Importance.BEST_PRACTICE_SUGGESTION, neurodata_type=Subject)
 def check_subject_sex(subject: Subject):
-    """Check if the subject sex has been specified, if one exists."""
+    """Check that the subject sex has been correctly specified."""
     if subject and not subject.sex:
         return InspectorMessage(message="Subject.sex is missing.")
-    elif subject.sex not in ("M", "F", "O", "U"):
+    if subject.species in ("Caenorhabditis elegens", "C. elegens") and subject.sex not in ("XO", "XX"):
```
Comment (CodyCBakerPhD) [inline src/nwbinspector/checks/nwbfile_metadata.py]: Just to confirm, the `C. elegens` pattern is there simply to inform the user of this additional sex option, not to say that `C. elegens` is an acceptable string for the latin binomial (which is a separate check)?

**Comment by CodyCBakerPhD** [comment]
XX and XO appear to be officially designated for this species, so yes perfectly fine by me

**Comment by CodyCBakerPhD** [comment]
The other life cycle stuff is a part of the ongoing discussion for handling age: https://github.com/NeurodataWithoutBorders/nwbinspector/issues/209

File: src/nwbinspector/checks/nwbfile_metadata.py (hunk 1 above)
Comment (bendichter) [reply src/nwbinspector/checks/nwbfile_metadata.py]: yes, that's the intention

File: docs/best_practices/nwbfile_metadata.rst [hunk 2]
```diff
@@ -195,6 +195,8 @@ Sex
 The ``sex`` of the :ref:`nwb-schema:sec-Subject` should be specified as a single upper-case character among the
 following four possibilities: "M" (male), "F" (female), "U" (unknown), or "O" (other, for asexual species).
 
+C. elegens are an exception to this rule. For C. elegens, the sex should either be "XO" (male) or "XX" (hermaphrodite).
```
Comment (rly) [inline docs/best_practices/nwbfile_metadata.rst]:
```suggestion
C. elegans are an exception to this rule. For C. elegans, the sex should either be "XO" (male) or "XX" (hermaphrodite).
```

File: src/nwbinspector/checks/nwbfile_metadata.py (hunk 1 above)
Comment (rly) [inline src/nwbinspector/checks/nwbfile_metadata.py]:
```suggestion
    if subject.species in ("Caenorhabditis elegans", "C. elegans") and subject.sex not in ("XO", "XX"):
```

File: src/nwbinspector/checks/nwbfile_metadata.py [hunk 3]
```diff
@@ -197,10 +197,12 @@ def check_subject_id_exists(subject: Subject):
...
+        return InspectorMessage(message="For C. elegens, Subject.sex should be 'XO' or 'XX'.")
```
Comment (rly) [inline src/nwbinspector/checks/nwbfile_metadata.py]:
```suggestion
        return InspectorMessage(message="For C. elegans, Subject.sex should be 'XO' or 'XX'.")
```

File: src/nwbinspector/checks/nwbfile_metadata.py [hunk 4]
```diff
@@ -197,10 +197,12 @@ def check_subject_id_exists(subject: Subject):
...
 @register_check(importance=Importance.BEST_PRACTICE_SUGGESTION, neurodata_type=Subject)
 def check_subject_sex(subject: Subject):
-    """Check if the subject sex has been specified, if one exists."""
+    """Check that the subject sex has been correctly specified."""
     if subject and not subject.sex:
         return InspectorMessage(message="Subject.sex is missing.")
-    elif subject.sex not in ("M", "F", "O", "U"):
+    if subject.species in ("Caenorhabditis elegans", "C. elegans") and subject.sex not in ("XO", "XX"):
+        return InspectorMessage(message="For C. elegans, Subject.sex should be 'XO' or 'XX'.")
+    if subject.sex not in ("M", "F", "O", "U"):
```
Comment (rly) [inline src/nwbinspector/checks/nwbfile_metadata.py]:
```suggestion
    elif subject.sex not in ("M", "F", "O", "U"):
```

Comment (bendichter) [reply src/nwbinspector/checks/nwbfile_metadata.py]: why change to elif? Won't this only come into play when the above condition is false?

**Comment by CodyCBakerPhD** [comment]
Just need a couple tests and this is good to go

File: src/nwbinspector/checks/nwbfile_metadata.py (hunk 4 above)
Comment (rly) [reply src/nwbinspector/checks/nwbfile_metadata.py]: I think if subject.sex is "XX", the second if will trigger and return the InspectorMessage

Comment (rly) [reply src/nwbinspector/checks/nwbfile_metadata.py]: Ah, you're right - the elif isn't sufficient either.

Comment (rly) [inline src/nwbinspector/checks/nwbfile_metadata.py]:
```suggestion
    if subject.species in ("Caenorhabditis elegans", "C. elegans"):
        if subject.sex not in ("XO", "XX"):
            return InspectorMessage(message="For C. elegans, Subject.sex should be 'XO' or 'XX'.")
    elif subject.sex not in ("M", "F", "O", "U"):
```

Comment (CodyCBakerPhD) [inline src/nwbinspector/checks/nwbfile_metadata.py]:
```suggestion
    elegans_references = ("Caenorhabditis elegans", "C. elegans")
    if subject.species in elegans_references  and subject.sex not in ("XO", "XX"):
        return InspectorMessage(message="For C. elegans, Subject.sex should be 'XO' or 'XX'.")
    if subject.species not in elegans_references and subject.sex not in ("M", "F", "O", "U"):
```

Flatter is better than nested, according to Python zen (i.e., reduce cyclomatic complexity)

If we ever need more special cases for other species in the far future, we can start using switch statements in Python 3.10

**Comment by CodyCBakerPhD** [comment]
@rly I remember you mentioning something about the reasons the DANDI tests are failing - can you confirm there is no easy way to fix the CI here?

**Comment by rly** [comment]
The tests fail because dandi raises DeprecationWarnings into errors. We have a fix in the dev branch of PyNWB that we haven't released yet. I'm getting to it...

**Comment by CodyCBakerPhD** [comment]
> The tests fail because dandi raises DeprecationWarnings into errors. We have a fix in the dev branch of PyNWB that we haven't released yet. I'm getting to it...

Thanks for confirming, no huge rush on my end I just wanted to know for sure

File: src/nwbinspector/checks/nwbfile_metadata.py (hunk 4 above)
Comment (CodyCBakerPhD) [reply src/nwbinspector/checks/nwbfile_metadata.py]: Even what I suggested here is more logic than necessary; added new approach with helper functions that will scale well with adding more special cases to future species, if necessary.

**Comment by CodyCBakerPhD** [comment]
@rly Good to merge?

**Review by CodyCBakerPhD — APPROVED** [review]

**Pull request merged**